Issue #600: Implement DATEDIFF

hawkfish opened the pull request:
This function is an implementation of the "period-crossing" [date difference function](https://docs.microsoft.com/en-us/sql/t-sql/functions/datediff-transact-sql?view=sql-server-ver15) from T-SQL.

Commit:
Issue #600: Implement DATE_DIFF

Original macro version.

Commit:
Issue #600: Implement DATEDIFF

C++ implementation and tests.

Commit:
Issue #600: Implement DATEDIFF

Actually delete old test file.

Mytherin commented on src/function/scalar/date/date_diff.cpp:
```diff
@@ -0,0 +1,333 @@
+	struct MonthOperator {
+		template <class TA, class TB, class TR>
+		static inline TR Operation(TA startdate, TB enddate) {
+			return (Date::ExtractYear(enddate) * 12 + Date::ExtractMonth(enddate) - 1) -
```
Can we use Date::Convert here? ExtractMonth calls Date::Convert internally.

hawkfish replied:
Duh, one call instead of two. You would think I've never implement this function before...

Mytherin commented on src/function/scalar/date/date_diff.cpp:
```diff
@@ -0,0 +1,333 @@
+	struct QuarterOperator {
+		template <class TA, class TB, class TR>
+		static inline TR Operation(TA startdate, TB enddate) {
+			return (Date::ExtractYear(enddate) * 12 + Date::ExtractMonth(enddate) - 1) / Interval::MONTHS_PER_QUARTER -
```
Same here!

Mytherin commented on src/function/scalar/date/date_diff.cpp:
```diff
@@ -0,0 +1,333 @@
+// This function is an implementation of the "period-crossing" date difference function from T-SQL
+// https://docs.microsoft.com/en-us/sql/t-sql/functions/datediff-transact-sql?view=sql-server-ver15
+struct DateDiff {
+	static constexpr const int32_t SECS_PER_WEEK = Interval::SECS_PER_DAY * Interval::DAYS_PER_WEEK;
```
These constants are mostly gathered in the `Interval` class right now, perhaps nice to group them up either here or there? Or in another place?

hawkfish replied:
Looks like some of them got added already!

Mytherin commented on src/function/scalar/date/date_diff.cpp:
```diff
@@ -0,0 +1,333 @@
+	auto &startdate_arg = args.data[1];
+	auto &enddate_arg = args.data[2];
+
+	TernaryExecutor::Execute<string_t, T, T, int64_t>(part_arg, startdate_arg, enddate_arg, result, args.size(),
```
Perhaps add an optimization for the (common?) scenario where part_arg is a constant?

hawkfish replied:
I was thinking of doing this on a second pass, but since I forgot to push, maybe I should just go ahead...

Mytherin reviewed (commented):
Thanks for the PR! Looks great. Some comments:

Commit:
Merge branch 'master' into hawkfish-datediff

Commit:
Issue #600: Implement DATEDIFF

Use single Date::Convert call instead of two ExtractXXX calls.
Move constants to interval.hpp

Commit:
Issue #600: Implement DATEDIFF

Special case constant date part as that is a common case.

Mytherin commented:
Thanks, looks great!
